Avm

mellowCS opened the pull request.

mellowCS committed:
Added the AVM spider to the spiders folder.

mellowCS committed:
Improved readability of parse_product

mellowCS committed:
avm.py: added download functionality using scrapy's file pipline
	- in settings.py the pipeline had to be activated and a valid file path had to be added
	- in items.py, fields for the file download were added

	removed start_requests() method and renamed 'urls' to 'start_urls' as the default function gets called
	automatically when the name 'start_urls' is found within the class.

	refactored the list comprehension into its own link_extractor() method for readability purposes

added test_avm.py: still have to figure out how to write proper tests for scrapy spiders

added html_files: for testing purposes

mellowCS committed:
Merge branch 'master' of https://github.com/mellowCS/FirmwareScraper into avm_spider

mellowCS committed:
Added creation of json file with meta data, including device class, release date etc., to the avm spider.

mellowCS committed:
added tests for parse, link_extractor and date_extractor

mellowCS committed:
added more test functionality

mellowCS committed:
added set_device_class method to avm spider and corresponding test
finished test for prepare_item_download()

mellowCS committed:
set new download timeout of 320 seconds

mellowCS committed:
added date converter and version extractor to avm crawler
added test for date converter

mellowCS committed:
removed __pycache__

mellowCS committed:
removed __pycache__

mellowCS committed:
added test for version extractor

mellowCS committed:
added mock_classes.py
added permutations method for improved version extraction and added corresponding test
moved from pytest_test mock to pytest's internal  monkeypatch functionality for mocking

mellowCS committed:
-a

mellowCS committed:
fixed wrong type hinting for returns
added cb_kwargs parameter for MockRequest

mellowCS committed:
renamed the 'prefix' parameter of the extract_links method to 'ignore' for
better readability as the parameter's value(s) shall be ignored when extracting links

dorpvom commented on firmware/pipelines.py:
```diff
@@ -5,7 +5,14 @@
+from scrapy.pipelines.files import FilesPipeline
+
 
 class FirmwarePipeline(object):
```
See other PR. Unused code

dorpvom commented on firmware/spiders/avm.py:
```diff
@@ -0,0 +1,111 @@
+
+
+class AvmSpider(Spider):
+    handle_httpstatus_list = [404]
```
See other PR. Check if spider globals can be named upper case (as in PEP8)

dorpvom commented on firmware/spiders/avm.py:
```diff
@@ -0,0 +1,111 @@
+
+    @staticmethod
+    def map_device_class(product: str) -> str:
+        if product.startswith(('fritzrepeater', 'fritzwlan-repeater')):
```
Refactor. You can replace `elif` by `if` as long as there is a `return` in each statement. Also `else` is unnecessary

dorpvom commented on firmware/spiders/avm.py:
```diff
@@ -0,0 +1,111 @@
+        if 'fritz.powerline' in firmware:
+            for hardware_number in self.permutations(array=product_specifier.split('-')[1:], prefix='', index=0):
+                matches = search(r'(?:' + r''.join(hardware_number.upper()) + r')_(.*).image', firmware)
+                if matches:
```
What happens if that statement is never true? Maybe you should define a fallback value. Otherwise this can return a `None` in some (edge) cases.

dorpvom commented on firmware/spiders/avm.py:
```diff
@@ -0,0 +1,111 @@
+        else:
+            return search(r'FRITZ\.(Box|Powerline|Repeater)_(\w+)(\.(\w{2}-)+\w{2}\.)?([-\.])?(.*)\.image', firmware).group(6)
+
+    def permutations(self, array: list, prefix: str, index: int) -> Generator[str, None, None]:
```
rename to sth. like get_permutations

dorpvom commented on firmware/tests/test_avm.py:
```diff
@@ -0,0 +1,157 @@
+        assert request.url == expected
+
+
+@pytest.mark.parametrize('response, expected', [(MockResponse(url='/fritzbox/fritzbox-1234/', body=LOCATION_PAGE), ['/fritzbox/fritzbox-1234/deutschland/', '/fritzbox/fritzbox-1234/other/']),
```
This can be better "line breaked". Put each `MockResponse` in a line on it's own.

dorpvom commented on firmware/tests/test_avm.py:
```diff
@@ -0,0 +1,157 @@
+import pytest
+
+import avm
```
Tests don't run on my machine.

Using these imports it works for me:

```python
from firmware.spiders import avm
from firmware.tests.mock_classes import MockResponse, MockRequest
```

dorpvom commented on firmware/spiders/avm.py:
```diff
@@ -0,0 +1,111 @@
+        loader.add_value('release_date', meta_data['release_date'])
+        yield loader.load_item()
+
+    def prepare_meta_data(self, response: Response, device_name: str, release_date: str, file_url: str) -> dict:
```
This function is not tested

dorpvom reviewed (changes requested):
Please attend to the requested changes

mellowCS committed:
swapped out the custom convert_date function with a build-in function
changed name of permutations to get_permutations
added a test for prepare_meta_data
fixed some indentation issues

dorpvom committed:
Merge branch 'avm' of https://github.com/mellowCS/FirmwareScraper into avm

dorpvom committed:
Merge branch 'master' of https://github.com/mellowCS/FirmwareScraper into avm

dorpvom committed:
added test, refactored code, took care of possible bug
